src/keys/bip39/mod.rs

```diff
@@ -0,0 +1,892 @@
+            }
+        }
+
+        let hash = sha256::Hash::hash(entropy);
+        // Possible size for check is >= 4 and <= 8
+        for j in 0..check_bits {
+            // Append check at the end of entropy bits
+            seed[seed_bits - check_bits + j] = hash[0] & (1 << (7 - j));
+        }
+
+        if seed.len() != seed_bits {
+            return Err(Error::InvalidChecksum(check_bits));
+        }
+
+        let mut words: Vec<u16> = vec![0; word_count];
+        for i in 0..word_count {
+            let mut index = 0;
+            for j in 0..BITS_PER_WORD {
+                if seed[i * BITS_PER_WORD + j] > 0 {
+                    index += 1 << (10 - j);
+                }
+            }
+            words[i] = index;
+        }
+
+        Ok(Mnemonic { language, words })
+    }
+
+    /// Convert a mnemonic to a seed with an optional passphrase
+    fn to_seed(&self, passphrase: Option<String>) -> [u8; 64] {
+        const PBKDF2_ITERATIONS: u32 = 2048;
+        const PBKDF2_BYTES: usize = 64;
+
+        let password = self
+            .word_iter()
+            .collect::<Vec<&str>>()
+            .join(" ")
+            .nfkd()
+            .collect::<String>();
+        let salt = ("mnemonic".to_string() + &passphrase.unwrap_or_else(|| "".to_string()))
+            .nfkd()
+            .collect::<String>();
+
+        let mut seed = [0u8; PBKDF2_BYTES];
+        pbkdf2::pbkdf2::<Hmac<Sha512>>(
+            password.as_bytes(),
+            salt.as_bytes(),
+            PBKDF2_ITERATIONS,
+            &mut seed,
+        );
+
+        seed
+    }
+
+    /// Convert a vec of word indices to an iterator of the mnemonic words
+    fn word_iter(&self) -> impl Iterator<Item = &str> + '_ {
+        let wordlist = self.language.wordlist();
+        self.words.iter().map(move |&w| wordlist[w as usize])
+    }
+}
+
+impl fmt::Display for Mnemonic {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        for (i, word) in self.word_iter().enumerate() {
+            if i > 0 {
+                f.write_str(" ")?;
+            }
+            f.write_str(word)?;
+        }
+        Ok(())
+    }
+}
+
+fn set_valid_on_any_network<Ctx: ScriptContext>(
+    descriptor_key: DescriptorKey<Ctx>,
+) -> DescriptorKey<Ctx> {
+    // We have to pick one network to build the xprv, but since the bip39 standard doesn't
+    // encode the network, the xprv we create is actually valid everywhere. So we override the
+    // valid networks with `any_network()`.
+    descriptor_key.override_valid_networks(any_network())
+}
+
+/// Type for a BIP39 mnemonic with an optional passphrase
+pub type MnemonicWithPassphrase = (Mnemonic, Option<String>);
+
+#[cfg_attr(docsrs, doc(cfg(feature = "keys-bip39")))]
+impl<Ctx: ScriptContext> DerivableKey<Ctx> for Seed {
+    fn into_extended_key(self) -> Result<ExtendedKey<Ctx>, KeyError> {
+        Ok(bip32::ExtendedPrivKey::new_master(Network::Bitcoin, &self[..])?.into())
```

Comment:
Would it make sense to implement `DerivableKey` for `(bip39::Seed, Network)` instead of for `bip39::Seed` directly. This can apply for the associated structures as well.

I'm suggesting this because some people may want to derive testnet keys from the Mnemonic/Seed, and this hard-coding may lead to some confusion.